File: App/Resources/AlgorithmDetails/guesssort/guesssort.py
def sort(arr):
    n = len(arr)
    loops = [0] * n
    indexes = [0] * n

    def is_valid():
        total = 0
        for i in range(n):
            for j in range(n):
                if loops[i] == loops[j]:
                    total += 1
        for i in range(n):
            for j in range(n):
                if i < j and arr[loops[i]] > arr[loops[j]]:
                    total += 1
                elif i > j and arr[loops[i]] < arr[loops[j]]:
                    total += 1
        return total == n

    while True:
        if is_valid():
            indexes[:] = loops
        pos = 0
        while pos < n:
            if loops[pos] < n - 1:
                loops[pos] += 1
                break
            else:
                loops[pos] = 0
                pos += 1
        if pos == n:
            break

    original = arr[:]
    for i in range(n):
        arr[i] = original[indexes[i]]


array = [0, 39, 21, 14]
sort(array)
print(array)

File: App/Resources/AlgorithmDetails/guesssort/guesssort.js
function sort(arr) {
    const n = arr.length;
    const loops = new Array(n).fill(0);
    const indexes = new Array(n).fill(0);

    function isValid() {
        let total = 0;
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (loops[i] === loops[j]) {
                    total += 1;
                }
            }
        }
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (i < j && arr[loops[i]] > arr[loops[j]]) {
                    total += 1;
                } else if (i > j && arr[loops[i]] < arr[loops[j]]) {
                    total += 1;
                }
            }
        }
        return total === n;
    }

    while (true) {
        if (isValid()) {
            for (let i = 0; i < n; i++) {
                indexes[i] = loops[i];
            }
        }
        let pos = 0;
        while (pos < n) {
            if (loops[pos] < n - 1) {
                loops[pos] += 1;
                break;
            } else {
                loops[pos] = 0;
                pos += 1;
            }
        }
        if (pos === n) {
            break;
        }
    }

    const original = arr.slice();
    for (let i = 0; i < n; i++) {
        arr[i] = original[indexes[i]];
    }
}

var array = [0, 39, 21, 14];
sort(array);
console.log("[" + array.join(", ") + "]");
